### PR TITLE
RIDEI-1391 

### PR DESCRIPTION
**Pull request by MunteanBeniamin**
Change the old implementation to match the new endpoints described in the [RIDEI-1391](https://jira.specialized.com/browse/RIDEI-1391) ticket. 

Note: calling the confirm endpoint for strava returns a failure message with a http redirect. -> this should be fixed by zone5 backed team.

### CODE REVIEW

File: Zone5.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1237,7 +1237,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = 31;
 				DEFINES_MODULE = YES;
-				DEVELOPMENT_TEAM = 986GGC5BDP;
+				DEVELOPMENT_TEAM = ZSK35JGVHZ;
```
Comment (jeremyTack) [inline Zone5.xcodeproj/project.pbxproj]: Should this be reverted?

File: Zone5/ThirdPartyConnections/ThirdPartyResponse.swift [hunk 2]
```diff
@@ -8,7 +8,8 @@
 
 import Foundation
 
-/// Response from ThirdPartyConnectionsView.setThirdPartyToken and removeThirdPartyToken
-public struct ThirdPartyResponse: Codable {
-	public let success: Bool
+/// Response from ThirdPartyConnectionsView.hasThirdPartyToken
+struct ThirdPartyResponse: Codable {
+    let type: String
+	let enabled: Bool
```
Comment (jeremyTack) [inline Zone5/ThirdPartyConnections/ThirdPartyResponse.swift:14]: It looks like your spacing is different than what was used for this file and `ThirdPartyConnectionsView`.

Comment (amonick12) [inline Zone5/ThirdPartyConnections/ThirdPartyResponse.swift:14]: indentation off here

File: Zone5.xcodeproj/project.pbxproj (hunk 1 above)
Comment (amonick12) [inline Zone5.xcodeproj/project.pbxproj]: Is this change necessary?

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 3]
```diff
@@ -9,10 +9,18 @@
...
+            self.uri = uri
+        }
+    }
 	private enum Endpoints: String, InternalRequestEndpoint {
-		case setThirdPartyConnection = "/rest/users/connections/api/v1/live_activities/set_third_party_token"
-		case hasThirdPartyConnection = "/rest/users/connections/api/v1/live_activities/has_third_party_token"
-		case removeThirdPartyConnection = "/rest/users/connections/api/v1/live_activities/delete_third_party_token"
+        case initializePairing = "/rest/users/connections/pair/"
+		case files = "/rest/files/"
+        case userConnections = "/rest/users/connections"
```
Comment (amonick12) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: indentation off here

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 4]
```diff
@@ -40,21 +48,44 @@ public class ThirdPartyConnectionsView: APIView {
 		return delete(endpoint, with: completion)
 	}
 
+    /// Initialize a connection for the current user for the given 3rd party type
+    /// - Parameters
+
```
Comment (amonick12) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: extra line here

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 5]
```diff
@@ -40,21 +48,44 @@ public class ThirdPartyConnectionsView: APIView {
...
+    }
+
 	/// Set an access token for the current user for the given 3rd party type
 	/// - Parameters
-	
+
 	@discardableResult
-	public func setThirdPartyToken(type: UserConnectionType, connection: ThirdPartyToken, completion: @escaping Zone5.ResultHandler<ThirdPartyResponse>) -> PendingRequest? {
-		return post(Endpoints.setThirdPartyConnection, parameters: queryParams(type), body: connection, with: completion)
+    public func setThirdPartyToken(type: UserConnectionType, parameters: URLEncodedBody?, completion: @escaping Zone5.ResultHandler<Zone5.VoidReply>) -> PendingRequest? {
```
Comment (amonick12) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: indentation off here

**Review by jeremyTack — CHANGES_REQUESTED** [review]

**Review by amonick12 — CHANGES_REQUESTED** [review]

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 6]
```diff
@@ -40,21 +48,44 @@ public class ThirdPartyConnectionsView: APIView {
...
+        let endpoint = ThirdPartyEndpoint(uri: Endpoints.files.rawValue + type.connectionName + "/confirm")
+        return get(endpoint, parameters: parameters, expectedType: Zone5.VoidReply.self, with: completion)
```
Comment (jeanhall) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: Take a look at other Endpoints in other views where urls requires string replacement formatting. The pattern used is to define the Endpoints url with a variable, e.g. for confirm it would be 
`case confirm = "/rest/files/{type}/confirm"` 
and call it here in the setThirdPartyToken func with Endpoints.confirm.replacingTokens(["type": type])

Do this with all the endpoints, it will clean this file up considerably, you won't the` private class ThirdPartyEndpoint: InternalRequestEndpoint` class and each Endpoints enum will map to a distinct server endpoint, which is what it is supposed to represent, rather than it just being a static string fragment.

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 7]
```diff
@@ -40,21 +48,44 @@ public class ThirdPartyConnectionsView: APIView {
...
+        }
 	}
 
 	@discardableResult
-	public func removeThirdPartyToken(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<ThirdPartyResponse>) -> PendingRequest? {
-		return post(Endpoints.removeThirdPartyConnection, parameters: queryParams(type), body: nil, with: completion)
+	public func removeThirdPartyToken(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<Bool>) -> PendingRequest? {
+        let endpoint = ThirdPartyEndpoint(uri: Endpoints.removeThirdPartyConnection.rawValue + type.connectionName)
+		return get(endpoint, with: completion)
 	}
```
Comment (jeanhall) [inline Zone5/Views/ThirdPartyConnectionsView.swift:77]: like above,  use replacingTokens

**Review by jeanhall — CHANGES_REQUESTED** [review]
* See inline comments - Endpoints enums should define the entire endpoint url using token replacement.
* Remove the change to developer team
* Fix indentation as per Aaron and Jermemy's comments
* Fix the Example App. It references the old endpoints and objects that have been removed so it will no longer compile. You can replace them with your new ones and then you can use the Example App to test your new endpoints against the server, just configure it to point to https://api-sp-staging.todaysplan.com.au with a valid clientId and secret and a valid user's credentials.

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 6 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Done

File: Zone5/ThirdPartyConnections/ThirdPartyResponse.swift (hunk 2 above)
Comment (MunteanBeniamin) [reply Zone5/ThirdPartyConnections/ThirdPartyResponse.swift:14]: Fixed

Comment (MunteanBeniamin) [reply Zone5/ThirdPartyConnections/ThirdPartyResponse.swift:14]: Fixed

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 3 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Fixed, this was a weird indentation, on my local changes was looking ok.

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 4 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Removed.

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 5 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Fixed

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 7 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift:77]: done

File: Zone5.xcodeproj/project.pbxproj (hunk 1 above)
Comment (MunteanBeniamin) [reply Zone5.xcodeproj/project.pbxproj]: Reverted

Comment (MunteanBeniamin) [reply Zone5.xcodeproj/project.pbxproj]: Yes, reverted

**Review by amonick12 — COMMENTED** [review]
I'm still seeing a lot of spacing inconsistencies.

**Comment by MunteanBeniamin** [comment]
> I'm still seeing a lot of spacing inconsistencies.

Check now, it should be good, I reindented all the changed classes, somehow the github showed a different indentation of what I see in Xcode.

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 8]
```diff
@@ -9,52 +9,76 @@
...
+        let endpoint = Endpoints.confirmConnection.replacingTokens(["connectionType": type.connectionName])
+        return get(endpoint, parameters: parameters, expectedType: Zone5.VoidReply.self, with: completion)
+    }
+
+    /// Checks if a connection type is enabled or not
+    /// - Parameters
+    @discardableResult
+    public func hasThirdPartyToken(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<Bool>) -> PendingRequest? {
+        return get(Endpoints.userConnections, parameters: queryParams(type), expectedType: [ThirdPartyResponse].self) { result in
+            switch result {
```
Comment (jeanhall) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: I don't think this server endpoint takes any query params: https://portal.zone5cloud.com/en/manager/api-documentation/whats-prod/api/api_reference_connections#get-user-connections

Also, since this does not need to call the func queryParams(type) in this class, I don't think this queryParams(type) func is used anywhere anymore and can be removed

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 9]
```diff
@@ -9,52 +9,76 @@
...
+        return delete(endpoint, with: completion)
+    }
+
+    /// Initialize a connection for the current user for the given 3rd party type
+    /// - Parameters
+    @discardableResult
+    public func initializeThirdPartyConnection(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<ThirdPartyInitializeResponse>) -> PendingRequest? {
+        let endpoint = Endpoints.initializePairing.replacingTokens(["connectionType": type.connectionName])
+        return post(endpoint, body: EmptyBody(), with: completion)
+    }
```
Comment (jeanhall) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: Do you need to send an empty body? Does it work if you just call `post(endpoint, body: nil, with: completion)`? body is an optional. You'll note that the old removeThirdPartyToken code did this. If you can just pass nil then the new EmptyBody struct is unnecessary.

File: Zone5/Views/ThirdPartyConnectionsView.swift [hunk 10]
```diff
@@ -9,52 +9,76 @@
...
+        case confirmConnection = "/rest/files/{connectionType}/confirm"
+        case userConnections = "/rest/users/connections"
+        case removeThirdPartyConnection = "/rest/users/connections/rem/{connectionType}"
+        case registerDeviceWithThirdParty = "/rest/users/scheduled/activities/api/v1/push_registrations"
+        case deregisterDeviceWithThirdParty = "/rest/users/scheduled/activities/api/v1/push_registrations/{token}"
+    }
+
+    private let serviceKey: String = "service_name"
+    internal func queryParams(_ serviceType: UserConnectionType) -> URLEncodedBody {
+        let queryParams: URLEncodedBody = [ serviceKey : "\(serviceType)" ]
```
Comment (jeanhall) [inline Zone5/Views/ThirdPartyConnectionsView.swift]: See below related comment about this func - I don't think that this func is used and can be removed.

**Review by jeanhall — CHANGES_REQUESTED** [review]
Looking better, a couple of things see inline comments AND the example app included in this project no longer compiles due to your changes. Please fix.

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 9 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Yes, this is needed. I tested with a nil and it returns a failure (decode error). James told me to send an empty body like this: "{}"

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 10 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Removed

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 8 above)
Comment (MunteanBeniamin) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Done. I don't have access to the documentation.

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 9 above)
Comment (jeanhall) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: ok

File: Zone5/Views/ThirdPartyConnectionsView.swift (hunk 8 above)
Comment (jeanhall) [reply Zone5/Views/ThirdPartyConnectionsView.swift]: Thanks. You should have access. You should be able to log in with an existing account (like what you would use for Ride), or just sign up a new account. Does that link not work?

**Review by amonick12 — APPROVED** [review]

**Review by jeanhall — APPROVED** [review]

**Review by jeremyTack — APPROVED** [review]

**Review by jellybeansoup — APPROVED** [review]
